### ISSUE TITLE
Is there an upper bound to how many experiment batches it is practical to train at once?

### ISSUE THREAD

**Issue by Munfred** [YosefLab/scvi-tools#838]
I was wondering if there is any clear limitation to how many separate batches it is practical to train at once?

For example, I typically train scVI on datasets with ~30 batches from multiple papers each with several experiments. In my case, I'm actually using all of C. elegans scRNseq datasets (there are only 5 papers out!), and I never had a problem. The really great advantage of using scVI to do the data integration for C. elegans is that I can train it once on all the public data, and then do DE on the cells of interest forever. I would like to be able to continue training scVI on all the data of the world rather than having to choose :)

However, I know that each additional batch is another neural network, and even while the minibatches allow indefinite scaling of the number of cells, I'm not sure how the memory scaling works as I increase the number of batches. Thus I imagine that at some point I'd have to reduce the number of batches (eg treating experiments from the same paper as one batch) in order to keep fitting it on the GPU memory.

Another strategy that I imagine would allow for scaling would be to train on CPU, which is a few times slower, but in which case I can easily run jobs with hundreds of GB of RAM available.

Thoughts?

**Comment by romain-lopez** [comment]
If you have a dataset with more than 100 batches I'd be happy to help. I think the "right" way to do this is to have a hierarchical model with a batch embedding learned along the cell embeddings and have the conventional neural nets use the embedding instead of the one-hot encoding.

**Comment by romain-lopez** [comment]
Once you do this, it can work again on CPU, GPU, anything you would like 💯

**Comment by adamgayoso** [comment]
I'll close this issue. @Munfred , if you'd like to make the first post on our new discourse server (discourse.scvi-tools.org) with this discussion that would be great! Probably goes in the models category if so.

**Comment by Munfred** [comment]
Yeah I wanna be the first poster! I'll do that now :D

**Comment by chansigit** [comment]
> If you have a dataset with more than 100 batches I'd be happy to help. I think the "right" way to do this is to have a hierarchical model with a batch embedding learned along the cell embeddings and have the conventional neural nets use the embedding instead of the one-hot encoding.

it's 2026 now, what would be the best solution now? is there any ready-to-use hierarchical models for us?

**Comment by ori-kron-wis** [comment]
The idea discussed in the forum (learning a hierarchical batch representation instead of one-hot encoding batches) remains mostly conceptual rather than a production model. See more info here: https://discourse.scverse.org/t/models-for-handling-hundreds-of-batches/24/3

However, MrVI (multi-resolution VI), designed specifically for many samples/batches, can effectively handle a large number of samples/batches without needing an explicit hierarchical batch model. If your batches have a nested structure, I would go with that first. 

I believe that with modern HW, compared to 2020, and perhaps with using our custom dataloaders, we can definitely train even a standard SCVI model on a large number of batches.

**Comment by chansigit** [comment]
> The idea discussed in the forum (learning a hierarchical batch representation instead of one-hot encoding batches) remains mostly conceptual rather than a production model. See more info here: https://discourse.scverse.org/t/models-for-handling-hundreds-of-batches/24/3论坛中讨论的想法（学习分层批次表示而不是对批次进行独热编码）仍然主要停留在概念阶段，而不是生产模型。更多信息请参见：https://discourse.scverse.org/t/models-for-handling-hundreds-of-batches/24/3
> 
> However, MrVI (multi-resolution VI), designed specifically for many samples/batches, can effectively handle a large number of samples/batches without needing an explicit hierarchical batch model. If your batches have a nested structure, I would go with that first.然而，MrVI（多分辨率变分推断）专门为处理许多样本/批次而设计，可以有效地处理大量样本/批次，而无需显式的分层批次模型。如果你的批次具有嵌套结构，我会首先选择该方法。
> 
> I believe that with modern HW, compared to 2020, and perhaps with using our custom dataloaders, we can definitely train even a standard SCVI model on a large number of batches.我认为，与 2020 年相比，利用现代硬件，也许还可以使用我们的自定义数据加载器，我们肯定可以在大量批次上训练甚至标准的 SCVI 模型。

Thanks for your advice. I'm so happy you could answer my question after years. I'll have a look at MrVI. :D

**Comment by canergen** [comment]
There's also embedding instead of one-hot for scVI (pass batch_representation='embedding' when setting up a model). This has been extensively evaluated for scPoli, and also performs well in terms of integration for scVI. However, in our experience the embeddings themselves are not meaningful representations to e.g. cluster diseased samples but contain in majority other nuisance variables.
